jwst/photom/photom.py

```diff
@@ -453,6 +513,98 @@ def calc_fgs(self, ftab):
             return 0.0
 
 
+    def calc_nrs_ifu_sens2d(self, area_data):
+
+        import math
+        import numpy as np
+        from .. assign_wcs import nirspec       # for NIRSpec IFU data
+
+        MICRONS_100 = 1.e-4                     # 100 microns, in meters
+
+        # Create empty 2D arrays for the wavelengths and pixel areas
+        wave2d = np.zeros_like(self.input.data)
+        area2d = np.ones_like(self.input.data)
+
+        # Create and initialize an array for the 2D dq map to be returned.
+        # initialize all pixels to NON_SCIENCE, because operations below
+        # only touch pixels within the bounding_box of each slice
+        dqmap = np.zeros_like(self.input.dq) + dqflags.pixel['NON_SCIENCE']
+
+        # Get the list of wcs's for the IFU slices
+        list_of_wcs = nirspec.nrs_ifu_wcs(self.input)
+
+        # Loop over the slices
+        for (k, ifu_wcs) in enumerate(list_of_wcs):
+
+            # Get the corners for the slice
+            xstart = ifu_wcs.bounding_box[0][0]
+            xstop = ifu_wcs.bounding_box[0][1]
+            ystart = ifu_wcs.bounding_box[1][0]
+            ystop = ifu_wcs.bounding_box[1][1]
+
+            log.debug("Slice %d: %g %g %g %g" % (k, xstart, xstop, ystart, ystop))
+
+            # Check for corner values off the edge
+            if xstart < -0.5:
+                log.debug("xstart from WCS bounding_box was %g" % xstart)
+                xstart = 0.
+            if ystart < -0.5:
+                log.debug("ystart from WCS bounding_box was %g" % ystart)
+                ystart = 0.
+            if xstop > 2047.5:
+                log.debug("xstop from WCS bounding_box was %g" % xstop)
+                xstop = 2047.
+            if ystop > 2047.5:
+                log.debug("ystop from WCS bounding_box was %g" % ystop)
+                ystop = 2047.
+
+            # Convert these to integers, and add one to the upper limits,
+            # because we want to use these as slice limits.
+            xstart = int(math.ceil(xstart))
+            xstop = int(math.floor(xstop)) + 1
+            ystart = int(math.ceil(ystart))
+            ystop = int(math.floor(ystop)) + 1
+
+            # Construct array indexes for pixels in this slice
+            dx = xstop - xstart
+            dy = ystop - ystart
+            ind = np.indices((dy, dx))
+            x = ind[1] + xstart
+            y = ind[0] + ystart
+
+            # Get the world coords for all pixels in this slice
+            coords = ifu_wcs(x, y)
+
+            # Pull out the wavelengths only
+            wl = coords[2]
+            nan_flag = np.isnan(wl)
+            good_flag = np.logical_not(nan_flag)
+            if wl[good_flag].max() < MICRONS_100:
+                log.info("Wavelengths in WCS table appear to be in meters")
```

Comment:
Early on the pipeline produced wavelengths in meters as this was how the models were made. Now wavelengths are in microns. Not sure if we need to keep checking this.